pyppeteer/domworld.py

```diff
@@ -0,0 +1,567 @@
+
+
+async def readFileAsync(path, file):
+    # TODO implement this
+    pass
+
+
+class DOMWorld(object):
+    def __init__(
+        self, frameManager: 'FrameManager', frame: 'Frame', timeoutSettings: TimeoutSettings,
+    ):
+        self._frameManager = frameManager
+        self._frame = frame
+        self._timeoutSettings = timeoutSettings
+        self._loop = self._frameManager._client._loop
+
+        self._documentFuture: Future['ElementHandle'] = None
+        self._contextFuture: Future['ExecutionContext'] = None
+        self._contextResolveCallback: Callable = None
+        self._setContext()
+
+        self._waitTasks = set()
+        self._detached = False
+
+        # Aliases for query methods:
+        self.J = self.querySelector
+        self.Jx = self.xpath
+        self.Jeval = self.querySelectorEval
+        self.JJ = self.querySelectorAll
+        self.JJeval = self.querySelectorAllEval
+
+    @property
+    def frame(self):
+        return self._frame
+
+    def _setContext(self, context: Optional['ExecutionContext'] = None):
+        if context:
+            self._contextResolveCallback(context)
+            self._contextResolveCallback = None
+            for waitTask in self._waitTasks:
+                waitTask.rerun()
+        else:
+            self._documentFuture = None
+            self._contextFuture = self._loop.create_future()
+            self._contextFuture.add_done_callback(lambda val: self._contextResolveCallback == val)
+
+    def _hasContext(self):
+        return not self._contextResolveCallback
+
+    def _detach(self):
+        self._detached = True
+        for task in self._waitTasks:
+            task.terminate(BrowserError('waitForFunctions failed: frame got detached.'))
+
+    @property
+    def executionContext(self) -> Awaitable['ExecutionContext']:
+        if self._detached:
+            raise BrowserError(
+                'Execution Context is not available in detached '
+                f'frame: {self._frame.url} (are you trying to evaluate?)'
+            )
+        return self._contextFuture
+
+    async def evaluateHandle(self, pageFunction: str, *args):
+        context = await self.executionContext
+        return context.evaluateHandle(pageFunction=pageFunction, *args)
+
+    async def evaluate(self, pageFunction: str, *args):
+        context = await self.executionContext
+        return context.evaluate(pageFunction=pageFunction, *args)
+
+    async def querySelector(self, selector: str):
+        document = await self._document
+        return await document.querySelector(selector)
+
+    @property
+    async def _document(self) -> 'ElementHandle':
+        if self._documentFuture:
+            return await self._documentFuture
+
+        context = await self.executionContext
+        document = await context.evaluateHandle('document')
+        self._documentFuture.set_result(document.asElement())
+        return await self._documentFuture
```

Comment:
This would be more analogous to the JS source I'm pretty sure. This will allow for `_document` to be called twice but only return the same future like (i think) the js authors intended. No references to `_document` will have to be changed as it returns an Awaitable
```suggestion
    @property
    def _document(self) -> Awaitable['ElementHandle']:
        if self._documentFuture:
            return await self._documentFuture

        async def _create_doc_future():
            context = await self.executionContext
            document = await context.evaluateHandle('document')
            return document.asElement()
        self._documentFuture = self._loop.create_task(_create_doc_future())
        return await self._documentFuture
```